### PR TITLE
RFC: Add separate CLUSTER grant

### PR DESCRIPTION
**Pull request by azat**
Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add separate CLUSTER grant (and `access_control_improvements.on_cluster_queries_require_cluster_grant` configuration directive, for backward compatibility, default to `false`)

In case you have different roles for the same user on multiple clusters,
ON CLUSTER query can help to overcome some limitations.

Consider the following example:
- cluster_with_data, dev_user (readonly=2)
- stage_cluster, dev_user (readonly=0)

So when you will execute the following query from stage_cluster, it will
be successfully executed, since ON CLUSTER queries has different system
profile:

    DROP DATABASE default ON CLUSTER cluster_with_data

This is not 100% safe, but at least something.

Note, that right now only ON CLUSTER query it self is supported, but
separate clusters are not (i.e. GRANT CLUSTER some_cluster_name TO
default), since right now grants sticked to database+.

Cc: @vitlibar

### CODE REVIEW

**Comment by vitlibar** [comment]
> > It seems this PR has slight compatibility issues because if the access is already set up it won't include that new grant, and thus after updating ClickHouse to a new version users will suddenly become unable to execute ON CLUSTER commands. I think we can solve that with using a new setting rbac_version which I'm adding in #35635
> 
> Indeed, but I though that this is OK, since you can do nasty things if user grants differs on the server on cluster. I'm fine to wait #35635, but can you elaborate on how you are seeing avoiding this issue with `rbac_version`?

I mean for every created or attached user we can check if it was created with old `rbac_version` and in that case just add `GRANT CLUSTER` to his grants

**Comment by vitlibar** [comment]
I've thought about these RBAC compatibility issues for a while and decided that probably the easiest way will be better, just because people will understand it better. So it seems we can add grants like this `CLUSTER` grant just by adding a new configuration options to the main configuration like this

```
<clickhouse>
    <access_control_improvements>
        <on_cluster_queries_require_cluster_grant>false</on_cluster_queries_require_cluster_grant>
    </access_control_improvements>
</clickhouse>
```

In my PR https://github.com/ClickHouse/ClickHouse/pull/36997 I used that approach. The default value for this setting will be false so there is no reason to worry for people who don't need this change.

**Comment by azat** [comment]
> So it seems we can add grants like this CLUSTER grant just by adding a new configuration options to the main configuration like this

OK, I've updated the patch.

File: tests/queries/0_stateless/02250_ON_CLUSTER_grant.reference [hunk 1]
```diff
@@ -0,0 +1,4 @@
+with_on_cluster_02250_ON_CLUSTER_grant_default
+localhost	9000	0		0	0
```
Comment (vitlibar) [inline tests/queries/0_stateless/02250_ON_CLUSTER_grant.reference:2]: Here is where the test failed, see https://s3.amazonaws.com/clickhouse-test-reports/35767/5920a535ff75e2484260a76807d3e496ea1508b7/stateless_tests__release__databasereplicated__actions__[1/2].html

I think we can redirect the output of the command `CREATE DATABASE` to `/dev/null` and only check stderr

File: src/Access/ContextAccess.cpp [hunk 2]
```diff
@@ -385,6 +385,9 @@ bool ContextAccess::checkAccessImplHelper(const AccessFlags & flags, const Args
     if (!getUser())
         return access_denied("User has been dropped", ErrorCodes::UNKNOWN_USER);
 
+    if (flags == AccessType::CLUSTER && !access_control->doesOnClusterQueriesRequireClusterGrant())
+        return access_granted();
```
Comment (vitlibar) [inline src/Access/ContextAccess.cpp:389]: `flags` is a bitmask and if `doesOnClusterQueriesRequireClusterGrant()` we need to ignore some bits and then check other bits as usual

File: tests/queries/0_stateless/02250_ON_CLUSTER_grant.reference (hunk 1 above)
Comment (azat) [reply tests/queries/0_stateless/02250_ON_CLUSTER_grant.reference:2]: >I think we can redirect the output of the command CREATE DATABASE ON CLUSTER to /dev/null and only check stderr

Yep, for this test it is not important.

File: src/Access/ContextAccess.cpp (hunk 2 above)
Comment (azat) [reply src/Access/ContextAccess.cpp]: Nice catch! Fixed.

**Review by vitlibar — APPROVED** [review]

**Pull request merged**